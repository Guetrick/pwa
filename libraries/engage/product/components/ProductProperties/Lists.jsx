import React from 'react';
import PropTypes from 'prop-types';
import { getSubgroupsFromProperties } from './helpers/getSubgroupsFromProperties';
import { getPropertiesWithoutSubgroup } from './helpers/getPropertiesWithoutSubgroup';
import { getPropertiesBySubgroup } from './helpers/getPropertiesBySubgroup';
import Group from './Group';
import Rows from './Rows';

/**
 * Renders lists of properties that are eventually grouped.
 * @param {Object} props The component props.
 * @return {JSX}
 */
const Lists = ({ properties }) => (
  <React.Fragment>
    <Rows properties={getPropertiesWithoutSubgroup(properties)} />
    {getSubgroupsFromProperties(properties).map(group => (
      <React.Fragment key={group}>
        <Group group={group} />
        <Rows properties={getPropertiesBySubgroup(properties, group)} />
      </React.Fragment>
    ))}
  </React.Fragment>
);

Lists.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default Lists;
