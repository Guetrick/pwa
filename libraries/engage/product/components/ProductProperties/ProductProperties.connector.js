import { connect } from 'react-redux';
import { makeGetProductProperties } from '../../selectors/product';

/**
 * @return {Function}
 */
function makeMapStateToProps() {
  const getProductProperties = makeGetProductProperties();

  return (state, props) => ({
    properties: getProductProperties(state, props),
  });
}

export default connect(makeMapStateToProps);
