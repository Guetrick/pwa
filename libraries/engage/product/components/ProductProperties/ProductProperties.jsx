import React from 'react';
import PropTypes from 'prop-types';
import { isBeta } from '../../../core';
import { ProductPropertiesSimple } from './ProductPropertiesSimple';
import { ProductPropertiesGrouped } from './ProductPropertiesGrouped';
import connect from './ProductProperties.connector';

/**
 * The product properties.
 * @returns {JSX}
 */
function ProductPropertiesContainer({ properties }) {
  if (!properties) {
    return null;
  }

  if (!isBeta()) { // TODO: Can be removed once the beta phase (product) is over!
    return <ProductPropertiesSimple properties={properties} />;
  }

  /*
    This feature is currently in BETA testing.
    It should only be used for approved BETA Client Projects
  */
  return (
    <ProductPropertiesGrouped properties={properties} />
  );
}

ProductPropertiesContainer.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesContainer.defaultProps = {
  properties: null,
};

export const ProductProperties = connect(ProductPropertiesContainer);
