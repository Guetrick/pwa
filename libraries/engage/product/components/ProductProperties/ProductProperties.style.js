import { css } from 'glamor';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';

const { variables: { gap } = {} } = themeConfig;

export const container = css({
  fontSize: '0.875rem',
  padding: `0 ${gap.big}px ${gap.big}px`,
  marginBottom: 12,
});

export const label = css({
  paddingBottom: (gap.small * 1.5),
});

export const tableCell = css({
  maxWidth: 100,
  padding: `${gap.xsmall * 0.5}px ${gap.small}px`,
  overflowWrap: 'break-word',
  ':first-of-type': {
    paddingLeft: 0,
  },
  ':last-of-type': {
    paddingRight: 0,
  },
});
