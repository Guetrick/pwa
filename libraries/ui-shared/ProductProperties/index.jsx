import React from 'react';
import PropTypes from 'prop-types';
import Ellipsis from '@shopgate/pwa-common/components/Ellipsis';

/**
 * The Product Properties component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const Properties = ({ truncate, properties }) => (
  <ul>
    {properties.map(({ label, value }) => (
      <li key={`${label}-${value}`}>
        {truncate &&
          <Ellipsis rows={truncate}>
            {label}: {value}
          </Ellipsis>
        }
        {!truncate && `${label}: ${value}`}
      </li>
    ))}
  </ul>
);

Properties.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.string,
  })),
  /** Truncate property to show N lines */
  truncate: PropTypes.number,
};

Properties.defaultProps = {
  properties: [],
  truncate: null,
};

export default Properties;
