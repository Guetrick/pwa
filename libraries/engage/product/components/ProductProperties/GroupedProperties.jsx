import React from 'react';
import PropTypes from 'prop-types';
import { Accordion } from '@shopgate/pwa-ui-material';
import Lists from './Lists';
import Wrapper from './Wrapper';
import { accordion } from './style';

/**
 * Renders the properties as groups.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const GroupedProperties = ({ properties, groups }) => groups
  .map(group => (
    <div key={group} className={accordion}>
      <Accordion renderLabel={() => group} testId={`product-properties-group-${group}`}>
        <Wrapper dense>
          <Lists
            properties={properties.filter(property => property.displayGroup === group)}
          />
        </Wrapper>
      </Accordion>
    </div>
  ));

GroupedProperties.propTypes = {
  groups: PropTypes.arrayOf(PropTypes.string).isRequired,
  properties: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default GroupedProperties;
