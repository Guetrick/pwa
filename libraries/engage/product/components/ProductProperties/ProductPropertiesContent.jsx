import React from 'react';
import PropTypes from 'prop-types';
import { isBeta } from '../../../core';
import { getGroupsFromProperties } from './helpers/getGroupsFromProperties';
import ProductPropertiesGrouped from './ProductPropertiesGrouped';
import ProductPropertiesWrapper from './ProductPropertiesWrapper';
import ProductPropertiesRows from './ProductPropertiesRows';
import { groupsContainer } from './ProductProperties.style';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
const ProductPropertiesContent = ({ properties }) => {
  if (!properties) {
    return null;
  }

  const groups = getGroupsFromProperties(properties);

  // Display the simple properties if no groups exist or if not in beta mode.
  if (!isBeta() || !groups || groups.length === 0) {
    return (
      <ProductPropertiesWrapper>
        <ProductPropertiesRows properties={properties} />
      </ProductPropertiesWrapper>
    );
  }

  /*
    This feature is currently in BETA testing.
    It should only be used for approved BETA Client Projects
  */
  return (
    <div className={groupsContainer}>
      <ProductPropertiesGrouped properties={properties} groups={groups} />
    </div>
  );
};

ProductPropertiesContent.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesContent.defaultProps = {
  properties: null,
};

export default ProductPropertiesContent;
