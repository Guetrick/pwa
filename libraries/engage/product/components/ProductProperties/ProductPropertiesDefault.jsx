import React from 'react';
import PropTypes from 'prop-types';
import { SurroundPortals } from '@shopgate/pwa-common/components';
import { PRODUCT_PROPERTIES } from '@shopgate/pwa-common-commerce/product/constants/Portals';
import { ProductPropertiesRows } from './ProductPropertiesRows';
import { container } from './ProductProperties.style';
import connect from './ProductPropertiesDefault.connector';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ProductPropertiesDefaultContainer({ properties }) {
  return (
    <SurroundPortals portalName={PRODUCT_PROPERTIES}>
      {(properties !== null && properties.length !== 0) && (
        <div className={container}>
          <table>
            <tbody>
              <ProductPropertiesRows properties={properties} />
            </tbody>
          </table>
        </div>
      )}
    </SurroundPortals>
  );
}

ProductPropertiesDefaultContainer.propTypes = {
  properties: PropTypes.arrayOf(PropTypes.shape()),
};

ProductPropertiesDefaultContainer.defaultProps = {
  properties: null,
};

export const ProductPropertiesDefault = connect(ProductPropertiesDefaultContainer);
