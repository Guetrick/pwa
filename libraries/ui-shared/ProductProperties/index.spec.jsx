import React from 'react';
import { mount } from 'enzyme';
import ProductProperties from './index';

describe('<ProductProperties />', () => {
  it('should not render when no properties are passed', () => {
    const wrapper = mount(<ProductProperties />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper).toBeEmptyRender();
  });

  it('should not render when an empty list of properties is passed', () => {
    const wrapper = mount(<ProductProperties properties={[]} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper).toBeEmptyRender();
  });

  it('should render as expected when properties are passed', () => {
    const properties = [{
      label: 'Label One',
      value: 'Value One',
    }, {
      label: 'Label Two',
      value: 'Value Two',
    }];

    const wrapper = mount(<ProductProperties properties={properties} />);
    expect(wrapper).toMatchSnapshot();
    const list = wrapper.find('li');
    expect(list.length).toBe(2);
    expect(list.at(0).text()).toBe(`${properties[0].label}: ${properties[0].value}`);
    expect(list.at(1).text()).toBe(`${properties[1].label}: ${properties[1].value}`);
  });
});
