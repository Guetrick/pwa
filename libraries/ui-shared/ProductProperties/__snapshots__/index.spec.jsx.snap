// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductProperties /> should not render when an empty list of properties is passed 1`] = `
<Properties
  lineClamp={null}
  properties={Array []}
/>
`;

exports[`<ProductProperties /> should not render when no properties are passed 1`] = `
<Properties
  lineClamp={null}
  properties={Array []}
/>
`;

exports[`<ProductProperties /> should render as expected when properties are passed 1`] = `
<Properties
  lineClamp={null}
  properties={
    Array [
      Object {
        "label": "Label One",
        "value": "Value One",
      },
      Object {
        "label": "Label Two",
        "value": "Value Two",
      },
    ]
  }
>
  <ul>
    <li
      key="Label One-Value One"
    >
      Label One: Value One
    </li>
    <li
      key="Label Two-Value Two"
    >
      Label Two: Value Two
    </li>
  </ul>
</Properties>
`;
